core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala

```diff
@@ -63,8 +65,48 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
   def ask[T: ClassTag](message: Any): Future[T] = ask(message, defaultAskTimeout)
 
   /**
-   * Send a message to the corresponding [[RpcEndpoint]] and get its result within a default
-   * timeout, or throw a SparkException if this fails even after the default number of retries.
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply]] and get its result within a
+   * default timeout, throw a SparkException if this fails.
+   *
+   * Note: this is a blocking action which may cost a lot of time,  so don't call it in a message
+   * loop of [[RpcEndpoint]].
+
+   * @param message the message to send
+   * @tparam T type of the reply message
+   * @return the reply message from the corresponding [[RpcEndpoint]]
+   */
+  def askWithBlocking[T: ClassTag](message: Any): T = askWithBlocking(message, defaultAskTimeout)
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply]] and get its result within a
+   * specified timeout, throw a SparkException if this fails.
+   *
+   * Note: this is a blocking action which may cost a lot of time, so don't call it in a message
+   * loop of [[RpcEndpoint]].
+   *
+   * @param message the message to send
+   * @param timeout the timeout duration
+   * @tparam T type of the reply message
+   * @return the reply message from the corresponding [[RpcEndpoint]]
+   */
+  def askWithBlocking[T: ClassTag](message: Any, timeout: RpcTimeout): T = {
+    try {
+      val future = ask[T](message, timeout)
+      val result = timeout.awaitResult(future)
+      if (result == null) {
+        throw new SparkException("RpcEndpoint returned null")
+      }
+      return result
+    } catch {
+      case NonFatal(e) =>
+        throw new SparkException(
```

Comment:
Isn't it better to just propagate the original exception? You can get the context from the stack trace.